keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,874 @@
+multiple levels of arrays above the IP/CIDR values
+(`networkpolicy.spec.egress[].to[].ipBlock.cidr`,
+`networkpolicy.spec.egress[].to[].ipBlock.except[]`,
+`networkpolicy.spec.ingress[].from[].ipBlock.cidr`,
+`networkpolicy.spec.ingress[].from[].ipBlock.except[]`,
+`endpoints.subsets[].addresses[].ip`,
+`endpoints.subsets[].notReadyAddresses[].ip`,
+`endpointslice.endpoints[].addresses[]`,
+`resourceclaim.status.devices[].networkData.ips[]`).
+
+In the current implementation, NetworkPolicy obeys an extended version
+of the simple array rule: any invalid CIDR value that appears in any
+field of the old object is allowed to appear in any field of the new
+object.
+
+On the other hand, Endpoints and EndpointSlice are treated
+differently, because (a) they are large enough that doing additional
+address-by-address validation on them could be noticeably slow, (b) in
+most cases, they are generated by controllers that only write out
+valid IPs anyway, (c) in the case of EndpointSlice, if we were going
+to allow moving bad IPs around within a slice without revalidation,
+then we ought to allow moving them between related slices too, which
+we can't really do.
+
+So for Endpoints and EndpointSlice, the rule will be that invalid IPs
+are only allowed to remain unfixed if the update leaves the entire
+top-level `.subsets` / `.endpoints` unchanged. So you can edit the
+labels or annotations without fixing invalid endpoint IPs, but you
+can't add a new IP while leaving existing invalid IPs in place.
+
+For ResourceClaim, since the field in question is part of an alpha
+feature, we will just update its validation to unconditionally require
+strict validation of all IP values.
+
+```
+<<[UNRESOLVED array ratcheting ]>>
+
+Is this behavior actually what we want, for all cases?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Ratcheting validation for immutable fields
+
+Four of the fields listed above are immutable:
+
+  - `pod.spec.dnsConfig.nameservers[]`
+  - `pod.spec.hostAliases[].ip`
+  - `service.spec.clusterIP`
+  - `service.spec.clusterIPs[]`
+
+It has been suggested that we should allow fixing invalid values in
+such fields. That is, you would be allowed to modify them if:
+
+  - the old value does not pass current validation rules, _and_
+  - the new value is the canonical representation of the old value.
+
+So given `clusterIP: 172.030.099.099`, you would be allowed to modify
+it to `clusterIP: 172.30.99.99`, but not to any other value. (For
+example, you could not modify it to `clusterIP: 172.30.99.099`,
+because while that is less wrong than the original value, it is still
+wrong, and not the canonical representation of that IP.)
+
+This allows people to fix bad values in existing objects, but adds
+complexity to validation, and potentially could cause problems with
+code that was expecting a more literal definition of "immutable".
+
+Also, in the case of the `Pod` fields, it shouldn't be too difficult
```

Comment:
Yeah, I don;t buy that we need to allow Pod updates (barring some new information).

Service is more approachable, I think.  IN particular, those fields are not ACTUALLY immutable, they just require hoop-jumping.  I think we can justify changing them (and in fact, we could argue that they should just be mutable, but that's another KEP entirely).

That said, this does seem lower prio than the rest.